File: src/snapshot/tx.py
"""Transaction List Builder interface for exec script"""
from enum import Enum

from gnosis.safe import Safe
from gnosis.safe.multi_send import MultiSendTx

from src.log import set_log
from src.multisend import build_multisend_from_data
from src.safe import encode_exec_transaction
from src.snapshot.delegate_registry import (
    encode_set_delegate,
    SAFE_DELEGATION_ID,
    encode_clear_delegate,
)

log = set_log(__name__)


class SnapshotCommand(Enum):
    """Enum containing all callable Snapshot contract methods"""

    SET_DELEGATE = "setDelegate"
    CLEAR_DELEGATE = "clearDelegate"


def transactions_for(
    parent: Safe, children: list[Safe], command: SnapshotCommand
) -> list[MultiSendTx]:
    """Builds transaction for given Snapshot command"""

    if command == SnapshotCommand.SET_DELEGATE:
        log.info(
            f"Setting delegation for namespace {SAFE_DELEGATION_ID} to parent {parent.address}"
        )
        return [
            build_multisend_from_data(
                safe=child,
                data=encode_exec_transaction(
                    child,
                    parent.address,
                    encode_set_delegate(SAFE_DELEGATION_ID, parent.address),
                ),
            )
            for child in children
        ]

    if command == SnapshotCommand.CLEAR_DELEGATE:
        return [
            build_multisend_from_data(
                safe=child,
                data=encode_exec_transaction(
                    child,
                    parent.address,
                    encode_clear_delegate(SAFE_DELEGATION_ID),
                ),
            )
            for child in children
        ]
    raise EnvironmentError(f"Invalid snapshot command: {command}")

File: rs/src/ops/snapshot/tx.rs
use alloy::dyn_abi::DynSolValue;
use alloy::primitives::{Address, U256};
use anyhow::Result;

use crate::multisend::MultiSendTx;
use crate::ops::snapshot::delegate_registry::{
    DelegationId, SAFE_DELEGATION_ID, delegation_address,
};
use crate::safe::tx::encode_contract_method;

/// Snapshot contract commands
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SnapshotCommand {
    SetDelegate,
    ClearDelegate,
}

impl SnapshotCommand {
    /// Get the method name for this command
    pub fn method_name(&self) -> &str {
        match self {
            SnapshotCommand::SetDelegate => "setDelegate",
            SnapshotCommand::ClearDelegate => "clearDelegate",
        }
    }
}

impl std::fmt::Display for SnapshotCommand {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.method_name())
    }
}

impl std::str::FromStr for SnapshotCommand {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        match s {
            "setDelegate" => Ok(SnapshotCommand::SetDelegate),
            "clearDelegate" => Ok(SnapshotCommand::ClearDelegate),
            _ => anyhow::bail!("Invalid snapshot command: {}", s),
        }
    }
}

/// Build transactions for a given snapshot command
pub fn transactions_for(
    parent_address: Address,
    children: &[Address],
    command: SnapshotCommand,
    delegate: Option<Address>,
) -> Result<Vec<MultiSendTx>> {
    let delegation_contract = delegation_address();
    let delegation_id = DelegationId::from_str(SAFE_DELEGATION_ID);

    let (params, method_signature) = match command {
        SnapshotCommand::SetDelegate => {
            let delegate_addr = delegate.unwrap_or(parent_address);
            log::info!(
                "Setting delegation for namespace {} to {:?}",
                SAFE_DELEGATION_ID,
                delegate_addr
            );

            // setDelegate(bytes32 id, address delegate)
            let delegation_bytes = delegation_id.as_bytes();
            (
                vec![
                    DynSolValue::FixedBytes(delegation_bytes.into(), 32),
                    DynSolValue::Address(delegate_addr),
                ],
                "setDelegate(bytes32,address)",
            )
        }
        SnapshotCommand::ClearDelegate => {
            log::info!("Clearing delegation for namespace {}", SAFE_DELEGATION_ID);

            // clearDelegate(bytes32 id)
            let delegation_bytes = delegation_id.as_bytes();
            (
                vec![DynSolValue::FixedBytes(delegation_bytes.into(), 32)],
                "clearDelegate(bytes32)",
            )
        }
    };

    let mut transactions = Vec::new();

    for child in children {
        let safe_tx = encode_contract_method(
            delegation_contract,
            method_signature,
            params.clone(),
            U256::ZERO,
        );

        transactions.push(safe_tx.as_multisend(*child, &parent_address)?);
    }

    Ok(transactions)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    #[test]
    fn test_snapshot_command_from_str() {
        assert_eq!(
            "setDelegate".parse::<SnapshotCommand>().unwrap(),
            SnapshotCommand::SetDelegate
        );
        assert_eq!(
            "clearDelegate".parse::<SnapshotCommand>().unwrap(),
            SnapshotCommand::ClearDelegate
        );
    }

    #[test]
    fn test_snapshot_command_method_name() {
        assert_eq!(SnapshotCommand::SetDelegate.method_name(), "setDelegate");
        assert_eq!(
            SnapshotCommand::ClearDelegate.method_name(),
            "clearDelegate"
        );
    }

    #[test]
    fn test_transactions_for_empty() {
        let parent = Address::from_str("0x1234567890123456789012345678901234567890").unwrap();
        let children: Vec<Address> = vec![];

        let txs = transactions_for(parent, &children, SnapshotCommand::SetDelegate, None).unwrap();

        assert_eq!(txs.len(), 0);
    }
}
